docs/smtchecker.rst

```diff
@@ -0,0 +1,733 @@
+.. _formal_verification:
+
+##################################
+SMTChecker and Formal Verification
+##################################
+
+Using formal verification, it is possible to perform an automated mathematical
+proof that your source code fulfills a certain formal specification.
+The specification is still formal (just as the source code), but usually much
+simpler.
+
+Note that formal verification itself can only help you understand the
+difference between what you did (the specification) and how you did it
+(the actual implementation). You still need to check whether the specification
+is what you wanted and that you did not miss any unintended effects of it.
+
+Solidity implements a formal verification approach based on SMT and Horn solving.
+The SMTChecker module automatically tries to prove that the code satisfies the
+specification given by ``require/assert`` statements. That is, it considers
+``require`` statements as assumptions and tries to prove that the conditions
+inside ``assert`` statements are always true.  If an assertion failure is
+found, a counterexample may be given to the user, showing how the assertion can
+be violated. If no warning is given by the SMTChecker for a property,
+it means that the property is safe.
+
+The other verification targets that the SMTChecker checks at compile time are:
+
+- Arithmetic underflow and overflow.
+- Division by zero.
+- Trivial conditions and unreachable code.
+- Popping an empty array.
+- Insufficient funds for a transfer.
+
+The potential warnings that the SMTChecker reports are:
+
+- ``<failing  property> happens here.``. This means that the SMTChecker proved that a certain property fails. A counterexample may be given, however in complex situations it may also not show a counterexample. This result may also be a false positive in certain cases, when the SMT encoding adds abstractions for Solidity code that is either hard or impossible to express.
+- ``<failing property> might happen here``. This means that the solver could not prove either case in the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard.
```

Comment:
```suggestion
- ``<failing property> might happen here``. This means that the solver could not prove either case within the given timeout. Since the result is unknown, the SMTChecker reports the potential failure for soundness. This may be solved by increasing the query timeout, but the problem might also simply be too hard.
```